common/sgx/eeid.c

```diff
@@ -0,0 +1,451 @@
+            return false;
+    return true;
+}
+
+oe_result_t verify_eeid(
+    const uint8_t* r_mrenclave,
+    const uint8_t* r_mrsigner,
+    uint16_t r_product_id,
+    uint32_t r_security_version,
+    uint64_t r_attributes,
+    const oe_eeid_t* eeid)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    if (!eeid)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (eeid->signature_size != 1808) // We only support SGX sigstructs for now.
+        OE_RAISE(OE_VERIFY_FAILED);
+
+    if (oe_get_current_logging_level() >= OE_LOG_LEVEL_WARNING)
+    {
+        char buf[2 * (sizeof(oe_eeid_t) + eeid->data_size) + 8];
+        OE_CHECK(oe_serialize_eeid(eeid, buf, sizeof(buf)));
+        printf("EEID:\n%s", buf);
+    }
+
+    // Compute expected mrenclave
+    OE_SHA256 cpt_mrenclave;
+    oe_replay_eeid_pages(eeid, &cpt_mrenclave, true);
+
+    // Extract reported mrenclave
+    OE_SHA256 reported_mrenclave;
+    uint8_t reported_mrsigner[OE_SIGNER_ID_SIZE];
+
+    memcpy(reported_mrenclave.buf, r_mrenclave, OE_SHA256_SIZE);
+    memcpy(reported_mrsigner, r_mrsigner, OE_SIGNER_ID_SIZE);
+
+    // Check recomputed mrenclave against reported mrenclave
+    if (memcmp(cpt_mrenclave.buf, reported_mrenclave.buf, OE_SHA256_SIZE) != 0)
+        OE_RAISE(OE_VERIFY_FAILED);
+
+    static const uint8_t debug_public_key[] = {
+        0xca, 0x9a, 0xd7, 0x33, 0x14, 0x48, 0x98, 0x0a, 0xa2, 0x88, 0x90,
+        0xce, 0x73, 0xe4, 0x33, 0x63, 0x83, 0x77, 0xf1, 0x79, 0xab, 0x44,
+        0x56, 0xb2, 0xfe, 0x23, 0x71, 0x93, 0x19, 0x3a, 0x8d, 0xa};
+
+    if (memcmp(debug_public_key, reported_mrsigner, OE_SIGNER_ID_SIZE) != 0)
+        OE_RAISE(OE_VERIFY_FAILED);
+
+    const sgx_sigstruct_t* sigstruct = (const sgx_sigstruct_t*)&eeid->signature;
+
+    // Compute and check base image hash
+    const uint8_t* base_mrenclave = sigstruct->enclavehash;
+    OE_SHA256 cpt_base_mrenclave;
+    oe_eeid_t tmp_eeid = *eeid;
+    // If we saved non-zero heap/stack sizes for the base image, we could add
+    // them here.
+    tmp_eeid.size_settings.num_heap_pages = 0;
+    tmp_eeid.size_settings.num_stack_pages = 0;
+    tmp_eeid.size_settings.num_tcs = 1;
+    oe_replay_eeid_pages(&tmp_eeid, &cpt_base_mrenclave, false);
+
+    if (memcmp(cpt_base_mrenclave.buf, base_mrenclave, OE_SHA256_SIZE) != 0)
+        OE_RAISE(OE_VERIFY_FAILED);
+
+    // Check other image properties have not changed
+    bool sigstruct_debug = sigstruct->attributes.flags & SGX_FLAGS_DEBUG;
+    bool reported_debug = r_attributes & OE_REPORT_ATTRIBUTES_DEBUG;
+
+    if (sigstruct_debug != reported_debug ||
+        sigstruct->isvprodid != r_product_id ||
+        sigstruct->isvsvn != r_security_version)
+        OE_RAISE(OE_VERIFY_FAILED);
+
+    // Check old signature (new signature has been checked above)
+    if (sigstruct_debug && is_zero(sigstruct->signature, OE_KEY_SIZE))
+        return OE_OK; // Unsigned debug image is ok?
+    else
+    {
+        unsigned char buf[sizeof(sgx_sigstruct_t)];
```

Comment:
This check is necessary in case somebody re-signs an image that has been tampered with (base images may be created by a different signer/person and a long time before the re-signing happens during enclave creation). But sure, I can move it into a separate function.

Reply:
(Done.)